### PR TITLE
Draft Posts

### PR DESCRIPTION
**Pull request by itays123**
- all new source code files are in typescript
- `src/services/drafts` can be used in all `contextType`s

### CODE REVIEW

**Comment by itays123** [comment]
It seems like the build is failing because of missing types in the vue composition api

File: src/services/drafts.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { api, getCallData } from './api'
+
+export function getDraft(contextType: string, contextId: string|null): Promise<any> {
```
Comment (davidmeirlevy) [inline src/services/drafts.ts:3]: use the crud service...
see other services for reference.

Comment (itays123) [reply src/services/drafts.ts:3]: The routes aren't exactly the same...

File: src/modules/posts/compositions/draft-post.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  function removeDraft() {
+    deleteDraft('post', postId);
+  }
+
+  return {
+    existingDraft,
+    saveAsDraft,
+    removeDraft
+  }
```
Comment (davidmeirlevy) [inline src/modules/posts/compositions/draft-post.ts:29]: You make a hook for drafts, so it's obvious those are drafts methods, so just return them as:
{ save, remove, draft }

File: src/services/drafts.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  return api.get('/api/drafts', { params: { contextType, contextId } }).then(getCallData);
+}
+
+export function setDraft(draftData: any) {
+  return api.put('/api/drafts', { data: draftData });
```
Comment (davidmeirlevy) [inline src/services/drafts.ts:8]: you don't chain the "getCallData" like in `getDraft`. that's why your build failed.

**Comment by itays123** [comment]
still failing. I'll look into it

**Comment by davidmeirlevy** [comment]
build locally before commit to see the errors:
https://g.codefresh.io/build/5f85d373f71e7398cd8f55fb?step=5f85d373f71e7398cd8f55fb&tab=output&logs=terminal

lint errors.

**Comment by itays123** [comment]
> lint errors.

CodeFresh doesn't let me view the console. what do I need to fix?

**Comment by davidmeirlevy** [comment]
run it locally, you'll get the lint errors in the terminal.

**Comment by itays123** [comment]
Build is successful!
Note that I deleted my entire custom hook and just worked on the existing hooks so all functionality is supposed to work. now the "debounce" feature is the only thing left to do.

**Review by davidmeirlevy — CHANGES_REQUESTED** [review]

File: src/services/drafts.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (itays123) [reply src/services/drafts.ts:3]: where is that folder?

File: src/modules/posts/compositions/unsaved-changes.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+function savePostDraft(postId: string|null, changes: any) {
+  setDraft({
+    contextType: 'post',
+    contextId: postId,
+    contextData: changes
+  })
+}
+
+export function useUnsavedChanges(postId: string|null = null, editedPost: any) {
```
Comment (davidmeirlevy) [inline src/modules/posts/compositions/unsaved-changes.ts:25]: The editedPost should be a Ref<any>, or else the watchEffect will not work..

**Review by davidmeirlevy — COMMENTED** [review]
I mentioned a tiny improvement.

Also, can you test the whole flow to see if it works? (Admin, front proxy, cli config, drafts service..)

**Comment by itays123** [comment]
> I mentioned a tiny improvement.
> 
> Also, can you test the whole flow to see if it works? (Admin, front proxy, cli config, drafts service..)

I'm having some trouble setting up a full greenpress environment. After I'll finish the denounce feature and the front proxy I'll try

**Comment by davidmeirlevy** [comment]
There's a lodash-debounce in this project. You can use it and see a reference on how to use it.

File: src/modules/posts/compositions/unsaved-changes.ts [hunk 5]
```diff
@@ -22,13 +23,11 @@ function savePostDraft(postId: string|null, changes: any) {
...
 
-export function useUnsavedChanges(postId: string|null = null, editedPost: any) {
+export function useUnsavedChanges(postId: string|null = null, editedPost: Ref<any>) {
   restoreDraft(postId, editedPost)
 
-  let timeoutId: number
   watchEffect(() => {
-    clearTimeout(timeoutId)
-    timeoutId = setTimeout(() => {
+    debounce(() => {
```
Comment (davidmeirlevy) [inline src/modules/posts/compositions/unsaved-changes.ts:30]: that's not how to use debounce.
you need to wrap savePostDraft with `debounce`, and use the debounced-function.

**Review by davidmeirlevy — CHANGES_REQUESTED** [review]

**Review by davidmeirlevy — APPROVED** [review]

**Pull request merged**